### DISCUSSION TITLE
Dynamic frames

### DISCUSSION THREAD

**Discussion by kelltom** [TomSchimansky/CustomTkinter#200] [Q&A]
Hello,

Firstly, I'd like to note that I am entirely new to Tkinter, so this may be a dumb question.

I'm making a desktop app that will control botting scripts for a video game (very evil of me, I know). Basically, I will be writing Python scripts to automate mouse/keyboard activity, and this app will be the one-stop shop to access all the scripts from a simple GUI interface. Each script is fairly unique, requiring certain user input before starting.  

**I'm wondering how I can make this into a single-page app with dynamically changing views/frames.**

The basic GUI template is something like this:
![1  Blank](https://user-images.githubusercontent.com/44652363/175162175-a391cc8b-3bd3-41fc-b556-8ee5c5374c4f.png)

The user opens the app and selects the script they'd like to run (assuming their game client is also open).
Upon clicking a script, the primary view should dynamically update according to the script, like so.
![2  What happens when you select a script](https://user-images.githubusercontent.com/44652363/175162362-480ed66e-3b1f-4a0a-9b4f-f312b83df6c8.png)

In simpler terms, for each script, I need to load a new pre-defined frame to replace the spot where the top-left rectangle is. The top right rectangle is probably going to be the same for all scripts, simple start/stop controls. The bottom rectangle will basically just print text information from the current running script and display it neatly - so it's constantly updating as the bot script is running (the bot needs to communicate with the GUI in real-time).

Before I get ahead of myself, is this achievable with CTk? If so, can anyone recommend a good design pattern that fits this nicely?

**Comment by felipetesc** [comment]
I believe you can achieve this.
I would: 
- for  each button click I would only care about selecting the correct frame which would call a gui to be rendered  inside options area, 
- by clicking at the same button I would select another code (inside a method, a class? you decide) responsible for things like play, pause, and stop (you can create a single class for that - it can be an abstrct class)
```python
from abc import ABC

# for each componenet implement this code

class FrameComponent(ABC):
	id = "invalid_id"
	gui = None
	config_properties = {}
	
	def play(self):
		raise Exception("Can't play, method unimplemented")

	def stop(self):
		raise Exception("Can't stop, method unimplemented")

	def pause(self):
   		raise Exception("Can't pause, method unimplemented")

   	def previous_step(self):
   		raise Exception("Can't come back, method unimplemented")

	def next_step(self):
   		raise Exception("Can't go forward, method unimplemented")


   	def create_ui(self, parent):
   		# create and return CTkFrame(parent)
   		# will allow you to further configs
   		# the config area can receive this CTkFrame to add to the config area
   		# save the states inside config_properties
   		raise Exception("Can't create ui, method unimplemented")

	def get_config_state(self):
		return FrameComponent.config_properties


	def run_script_on_console(self):
		# use subprocess
		# use process and some other related commands
		raise Exception("Can't pause, method unimplemented")
```

Readings:

For frame selection you can read:
- https://felipetesc.github.io/CtkDocs/#/multiple_frames
- https://github.com/TomSchimansky/CustomTkinter/discussions/99
- https://stackoverflow.com/questions/7546050/switch-between-two-frames-in-tkinter

For the terminal part you can read:
- https://pypi.org/project/tkterminal/
- https://stackoverflow.com/questions/63597533/how-to-add-a-console-to-a-tkinter-window
- https://stackoverflow.com/questions/7253448/how-to-embed-a-terminal-in-a-tkinter-application

For executing your code
- run python from python: https://stackoverflow.com/questions/701802/how-do-i-execute-a-string-containing-python-code-in-python
- subprocess discussion: https://docs.python.org/3/library/subprocess.html#module-subprocess
- execute c: https://stackoverflow.com/questions/4376397/executing-a-c-program-in-python

Hope it helps !

**Reply by kelltom** [reply]
Thanks for this! I had gone ahead and built some solutions prior to seeing this, but this really made me rethink what I was doing - in a good way. I used an MVC pattern, probably making my app more convoluted than it needs to be. Anyways, here's what I did.

Main view:
```python
class App(customtkinter.CTk):

    def __init__(self):
        super().__init__()

        # ==== imagine there's a bunch of configuration for the UI here - omitting for clarity ====

        # ============ frame_left (button menu) ============

        self.button_list = []
        self.btn_bot1= customtkinter.CTkButton(master=self.frame_left,
                                                      text="Bot1",
                                                      fg_color=("gray75", "gray30"),
                                                      command=lambda: self.show_frame("Bot1", self.btn_bot1))
        self.btn_bot1.grid(row=2, column=0, pady=10, padx=20)
        self.button_list.append(self.btn_bot1)

        # ============ frame_right (bot control panel) ============
        self.view_list = {}
        # Bot1
        self.view_list["Bot1"] = BotView(parent=self.frame_right)
        self.bot1_model= Bot1()
        self.bot1_controller= BotController(model=self.bot1_model, view=self.view_list["Bot1"])
        self.view_list["Bot1"].setup(controller=self.bot1_controller,
                                                   title=self.bot1_model.title,
                                                   description=self.bot1_model.description)

        # So if I wanted to add Bot2 to my app, I'd just add another button in the -frame_left- section,
        # and add another BotView to the view_list here.

    # ============ Script button handlers ============
    def hide_all_frames(self):
        # TODO: stop running threads
        # set all buttons to default color
        for button in self.button_list:
            button.config(fg_color=("gray75", "gray30"))
        for view in self.view_list.values():
            view.pack_forget()

    def show_frame(self, name, btn):
        self.hide_all_frames()
        btn.config(fg_color=btn.hover_color)
        self.view_list[name].pack(fill="both", expand=1)

    def on_closing(self, event=0):
        self.destroy()

    def start(self):
        self.mainloop()


if __name__ == "__main__":
    app = App()
    app.start()
```
This way, the *hide_all_frames()* and *show_frames()* functions will just pack/unpack views as they are selected.

The *BotView* contains everything you see on the right side of the UI below. The bot model class (*Bot1()* in my example above) stores the bot's config properties as well as running logic. The controller runs the bot's main loop on another thread so it doesn't lock up the UI and tells the view how it should update while the bot runs.

![bot](https://user-images.githubusercontent.com/44652363/177415562-2b417a66-2c99-4f7a-aafe-cc8e732f697e.png)

I pixelated some stuff to be more consistent with my example among other reasons.

Thought I'd add my own solution in case this helps anyone else out! Thanks for the help. I have absolutely zero experience with Tkinter outside of this so it's been fun!

**Reply by felipetesc** [reply]
This is just my personal perspective, but I prefer to avoid any repetition in UI code with loops if I can use direct access with array indexes. 

```python
def toggle_frame_by_id(self, frame_id):
        if App.frames[frame_id] is not None:
            if App.current is App.frames[frame_id]:
                App.current.pack_forget()
                App.current = None
            elif App.current is not None:
                App.current.pack_forget()
                App.current = App.frames[frame_id]
                App.current.pack(in_=self.right_side_panel, side=tkinter.TOP, fill=tkinter.BOTH, expand=True, padx=0, pady=0)
            else:
                App.current = App.frames[frame_id]
                App.current.pack(in_=self.right_side_panel, side=tkinter.TOP, fill=tkinter.BOTH, expand=True, padx=0, pady=0)
```

You can show, if that's possible, your project as a sample about what can be done with CTK.

Thanks.

**Reply by kelltom** [reply]
Your perspective is very correct, I suppose I was just being lazy lol. Thanks for the tip!

**Reply by avalon60** [reply]
MVC is a good pattern. Believe me, I have developed apps which get complex and MVC had been the way to go.

**Comment by jklouz** [comment]
.local/lib/python3.11/site-packages/customtkinter/windows/widgets/utility/utility_functions.py", line 18, in check_kwargs_empty
    raise ValueError(f"{list(kwargs_dict.keys())} are not supported arguments. Look at the documentation for supported arguments.")
ValueError: ['padx', 'pady'] are not supported arguments. Look at the documentation for supported arguments.

how to fix this?

**Reply by felipetesc** [reply]
code ?

**Reply by jklouz** [reply]
the code is your multiple frame example 2 code

**Reply by felipetesc** [reply]
The code used to work, but the api got changed:

Updated version below (tested on Windows 11,  customtkinter 5.2.0) 

```python
import tkinter
import customtkinter
from functools import partial

DARK_MODE = "dark"
customtkinter.set_appearance_mode(DARK_MODE)
customtkinter.set_default_color_theme("blue")


class App(customtkinter.CTk):

    frames = {}
    current = None
    bg = ""


    def __init__(self):
        super().__init__()
        self.bg = self.cget("fg_color")
        self.num_of_frames = 0
        # self.state('withdraw')
        self.title("Change Frames")

        # screen size
        self.geometry("800x600")

        # root!
        main_container = customtkinter.CTkFrame(self, corner_radius=8, fg_color=self.bg)
        main_container.pack(fill=tkinter.BOTH, expand=True, padx=8, pady=8)

        # left side panel -> for frame selection
        self.left_side_panel = customtkinter.CTkFrame(main_container, width=280, corner_radius=8, fg_color=self.bg)
        self.left_side_panel.pack(side=tkinter.LEFT, fill=tkinter.Y, expand=False, padx=18, pady=10)

        # right side panel -> to show the frame1 or frame 2, or ... frame + n where n <= 5
        self.right_side_panel = customtkinter.CTkFrame(main_container, corner_radius=8, fg_color="#212121")
        self.right_side_panel.pack(side=tkinter.LEFT, fill=tkinter.BOTH, expand=True, padx=0, pady=0)
        self.right_side_panel.configure(border_width = 1)
        self.right_side_panel.configure(border_color = "#323232")
        self.create_nav( self.left_side_panel, "frame1", self.color_by_id(1))
        self.create_nav( self.left_side_panel, "frame2", self.color_by_id(2))
        self.create_nav( self.left_side_panel, "frame3", self.color_by_id(3))
        self.create_nav( self.left_side_panel, "frame4", self.color_by_id(4))
        self.create_nav( self.left_side_panel, "frame5", self.color_by_id(5))


    def color_by_id(self, id):
        if id == 1: 
            return "blue"
        
        elif id == 2:
            return "green"
        
        elif id == 3:
            return "orange"
        
        elif id == 4:
            return "purple"
        
        elif id == 5:
            return "grey" 
        
        else:
            return "red"

    # button to select the correct frame
    def frame_selector_bt(self, parent, frame_id):
        
        # create frame 
        bt_frame = customtkinter.CTkButton(parent)
        # style frame
        bt_frame.configure(height = 40)
        # creates a text label
        bt_frame.configure(text = self.color_by_id(self.num_of_frames + 1)  )
        bt_frame.configure(command =  partial( self.toggle_frame_by_id,  "frame" + str(self.num_of_frames + 1) ) )
        # set layout
        bt_frame.grid(pady = 3, row=self.num_of_frames, column=0)
        # update state
        self.num_of_frames = self.num_of_frames + 1

    # create the frame
    def create_frame(self, frame_id, color):
        App.frames[frame_id] = customtkinter.CTkFrame(self, fg_color=self.cget("fg_color"))
        App.frames[frame_id].configure(corner_radius = 8)
        App.frames[frame_id].configure(fg_color = color)
        App.frames[frame_id].configure(border_width = 2)
        App.frames[frame_id].configure(border_color = "#323232")
        App.frames[frame_id].padx = 8

        bt_from_frame1 = customtkinter.CTkButton(App.frames[frame_id], text="Test "+ self.color_by_id(self.num_of_frames) , command=lambda:print("test " + color) )
        bt_from_frame1.place(relx=0.5, rely=0.5, anchor=tkinter.CENTER)

    # method to change frames
    def toggle_frame_by_id(self, frame_id):
        
        if App.frames[frame_id] is not None:
            if App.current is App.frames[frame_id]:
                App.current.pack_forget()
                App.current = None
            elif App.current is not None:
                App.current.pack_forget()
                App.current = App.frames[frame_id]
                App.current.pack(in_=self.right_side_panel, side=tkinter.TOP, fill=tkinter.BOTH, expand=True, padx=0, pady=0)
            else:
                App.current = App.frames[frame_id]
                App.current.pack(in_=self.right_side_panel, side=tkinter.TOP, fill=tkinter.BOTH, expand=True, padx=0, pady=0)

    # method to create a pair button selector and its related frame
    def create_nav(self, parent, frame_id, frame_color):
        self.frame_selector_bt(parent, frame_id)
        self.create_frame(frame_id, frame_color)


a = App()
a.mainloop()
```

![Untitled](https://github.com/TomSchimansky/CustomTkinter/assets/2336812/b2e413f1-ca28-4fde-81d9-9b6c7c66f7f9)

**Reply by jklouz** [reply]
Nice, thank you
